runtime/vm/OutOfLineINL_openj9_internal_foreign_abi_InternalDowncallHandler.cpp

```diff
@@ -216,6 +225,70 @@ OutOfLineINL_openj9_internal_foreign_abi_InternalDowncallHandler_initCifNativeTh
 	ffiTypeHelpers.freeStructFFIType(returnType);
 	goto done;
 }
+
+#if JAVA_SPEC_VERSION >= 22
+/* openj9.internal.foreign.abi.InternalDowncallHandler: private static native boolean isFfiProtoEnabled(); */
+VM_BytecodeAction
+OutOfLineINL_openj9_internal_foreign_abi_InternalDowncallHandler_isFfiProtoEnabled(J9VMThread *currentThread, J9Method *method)
+{
+	VM_BytecodeAction rc = EXECUTE_BYTECODE;
+	J9JavaVM *vm = currentThread->javaVM;
+	jboolean isFfiProtoOn = J9_ARE_ANY_BITS_SET(vm->extendedRuntimeFlags2, J9_EXTENDED_RUNTIME2_FFI_PROTO) ? JNI_TRUE : JNI_FALSE;
+
+	VM_OutOfLineINL_Helpers::returnSingle(currentThread, isFfiProtoOn, 0);
+	return rc;
+}
+
+/**
+ * openj9.internal.foreign.abi.InternalDowncallHandler: private native void resolveBoundHandle();
+ *
+ * @brief Resolve the bounded handle via MethodHandleResolver.ffiCallLinkCallerMethod() in advance
+ *        when calling into the interpreter via MH.linkToNative()
+ *
+ * @return void
+ */
+VM_BytecodeAction
+OutOfLineINL_openj9_internal_foreign_abi_InternalDowncallHandler_resolveBoundHandle(J9VMThread *currentThread, J9Method *method)
+{
+	VM_BytecodeAction rc = EXECUTE_BYTECODE;
+	j9object_t invokeCache = NULL;
+
+	jobject nativeInvoker = j9jni_createGlobalRef((JNIEnv *)currentThread, J9_JNI_UNWRAP_REFERENCE(currentThread->sp), JNI_FALSE);
```

Comment:
can this be a local ref instead?

Reply:
Will need to try out to see whether it works in the same way as `j9jni_createGlobalRef` in our case.

Reply:
Updated with `j9jni_createLocalRef` (used by `nativeMH`) and verified in tests.